### PR TITLE
Build the wgpu shader at runtime

### PR DESCRIPTION
**Pull request by khyperia**
This _dramatically_ improves the IDE experience when working on rust-gpu.

### CODE REVIEW

**Comment by VZout** [comment]
I think you should leave an option to pick between the two. in the end this is a example for other people and they might want to know they can compile at compile time.

**Comment by khyperia** [comment]
The android version does compile at compile time with this change, that's already an example.

**Review by VZout — APPROVED** [review]

**Review by termhn — APPROVED** [review]
Nice.

File: examples/runners/wgpu/src/lib.rs [hunk 1]
```diff
@@ -57,6 +57,46 @@ pub enum RustGPUShader {
...
+        let crate_path = [
+            Path::new(manifest_dir),
+            Path::new(".."),
+            Path::new(".."),
+            Path::new("shaders"),
+            Path::new(crate_name),
+        ]
+        .iter()
+        .copied()
+        .collect::<PathBuf>();
```
Comment (eddyb) [inline examples/runners/wgpu/src/lib.rs:81]: Isn't this shorter as a chain of `.join(...)` calls? Tho I suppose that may be less efficient hmpf.

Comment (khyperia) [reply examples/runners/wgpu/src/lib.rs:81]: yeah, probably, idk, I tend to over-optimize things that really I probably should just write the clean readable way :|

File: examples/runners/wgpu/build.rs [hunk 2]
```diff
@@ -1,15 +1,24 @@
...
+    // Always build on CI to make sure the shaders can still build
+    let is_on_ci = std::env::var("CI");
+    if target_os != "android" && target_arch != "wasm32" && is_on_ci.is_err() {
+        return Ok(());
+    }
+    let status = std::process::Command::new("cargo")
+        .args(["run", "--release", "-p", "example-runner-wgpu-builder"])
+        .stderr(std::process::Stdio::inherit())
+        .stdout(std::process::Stdio::inherit())
+        .status()?;
```
Comment (eddyb) [inline examples/runners/wgpu/build.rs:15]: Doesn't this deadlock Cargo without passing a different `--target-dir`? Have you tried building `example-runner-wgpu` with `--release` and `CI` being set? It should get stuck AFAICT (given e.g. `touch examples/runners/wgpu/builder/src/main.rs` to force a rebuild of it).

**Review by eddyb — APPROVED** [review]

**Pull request merged**